contracts-periphery/test/Gas.t.sol

```diff
@@ -0,0 +1,423 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "solmate/test/utils/mocks/MockERC20.sol";
+import "test/utils/DSTestPlus.sol";
+import "src/UmbraBatchSend.sol";
+
+interface UmbraToll {
+  function toll() external returns(uint256);
+}
+
+abstract contract UmbraBatchSendGasTest is DSTestPlus {
+  UmbraBatchSend router;
+  MockERC20 token;
+
+  address constant umbra = 0xFb2dc580Eed955B528407b4d36FfaFe3da685401;
+  address alice = address(0x202204);
+  address bob = address(0x202205);
+
+  uint256 ethBalance;
+  uint256 tokenBalance;
+
+  uint256 toll;
+  bytes32 pkx = "pkx";
+  bytes32 ciphertext = "ciphertext";
+
+  UmbraBatchSend.SendEth[] sendEth;
+  UmbraBatchSend.SendToken[] sendToken;
+
+
+  // Setup params for each function
+  enum Send {ETH, TOKEN, BOTH}
+  address payable[] addrs;
+  uint256 valueAmount; // {value: valueAmount} when calling Umbra
+  uint256 numOfAddrs;
+  uint256 etherAmount; // Ether amount to be sent per address
+  uint256 tokenAmount; // Token amount to be sent per address
+
+  function setUp() virtual public {
+    // Deploy Umbra at an arbitrary address, then place the resulting bytecode at the same address as the production deploys.
+    vm.etch(umbra, (deployCode("test/utils/Umbra.json", bytes(abi.encode(0, address(this), address(this))))).code);
```

Comment:
A few more changes here:
- It looks like there's still duplicate code you can move to DeployUmbraTest, namely the `constant` declarations for the users, pkx, and ciphertext
- The DeployUmbraTest has no setUp function—instead of calling `deployUmbra` in each test file's setup method, do that in DeployUmbraTest, and tests call `super.setUp()` to run that

Reply:
Made the changes here: https://github.com/ScopeLift/umbra-protocol/pull/338/commits/f1d5c52d16d2b85473ac20edf2ecc54aad1216ef